### PR TITLE
Update custom traces table with filters

### PR DESCRIPTION
**Pull request by ps48**
### Description

* The issue with original custom source table was that it says traces on top but shows data only from trace root spans. This is a problem cause it breaks users expectation of filtering traces across various fields like services and resource attributes. 

Sorting and Pagination push down won’t be part of this PR. 

Update custom traces table with filters added below:

* All Spans - All spans from all traces
* Traces - Aggregates all spans by traceId to show all traces
* Service Entry Spans - The spans that mark start of server-side processing (SPAN_KIND_SERVER)
* Trace Root Spans - The root spans which represent the starting point of a trace

Functional updates: 
* The trace table in custom source defaults to all spans
* The trace table mode is stored in session storage, to persist across tab refreshes and page changes

### Issues Resolved

* Default table view 
Screenshot 2024-09-20 at 9 26 28 AM

* Table filter options 
Screenshot 2024-09-20 at 9 26 38 AM

* Table showing trace root spans
Screenshot 2024-09-20 at 9 26 50 AM

* Table showing attributes
Screenshot 2024-09-20 at 9 27 00 AM

Demo Video

https://github.com/user-attachments/assets/6cc5060e-14a6-4ab6-9a2c-b31c16b86bd3

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: common/constants/trace_analytics.ts [hunk 1]
```diff
@@ -24,3 +24,35 @@ export const TRACE_ANALYTICS_DSL_ROUTE = '/api/observability/trace_analytics/que
...
+    'aria-describedby': 'The spans that mark start of server-side processing (SPAN_KIND_SERVER)',
+  },
+  {
+    label: TRACE_TABLE_TITLES.traces,
+    key: 'traces',
+    'aria-describedby': 'Aggregates all spans by traceId to show all traces',
+  },
+];
+
+export const TRACE_TABLE_TYPE_KEY = 'TraceAnalyticsTraceTableType';
```
Comment (joshuali925) [inline common/constants/trace_analytics.ts:58]: ideally the constants are in `common` only if they are shared between client (public) and server. if they are only used on one side they can go directly there

File: public/components/trace_analytics/components/traces/trace_table_helpers.tsx [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+      ),
+      align: 'right',
+      sortable: true,
+      truncateText: true,
+      render: (item) =>
+        item ? (
+          <EuiText>
+            <EuiToolTip content={item}>
+              <EuiText size="s">
+                {item.length < 36 ? item : <div title={item}>{truncate(item, { length: 36 })}</div>}
```
Comment (joshuali925) [inline public/components/trace_analytics/components/traces/trace_table_helpers.tsx:44]: use css to truncate if possible, like https://github.com/opensearch-project/dashboards-observability/blob/fc78c82795f5d8c138f424ccf6c7f96eca09e2c1/public/components/trace_analytics/index.scss#L75-L82

File: public/components/trace_analytics/index.scss [hunk 3]
```diff
@@ -103,6 +103,18 @@ th[data-test-subj^='tableHeaderCell_dashboard_latency_variance'] {
...
 .attributes-column-header {
   display: table-cell;
 }
+
+.tableModePopover {
+  width: 300px;
+}
+
+.popOverSelectableItem {
+  white-space: initial !important;
```
Comment (joshuali925) [inline public/components/trace_analytics/index.scss:119]: is !important necessary?

File: public/components/trace_analytics/components/traces/traces_custom_indices_table.tsx [hunk 4]
```diff
@@ -47,218 +45,117 @@ interface TracesLandingTableProps {
...
 
 export function TracesCustomIndicesTable(props: TracesLandingTableProps) {
   const { columnItems, items, refresh, mode, loading, getTraceViewUri, openTraceFlyout } = props;
   const [showAttributes, setShowAttributes] = useState(false);
+  const [isTitlePopoverOpen, setIsTitlePopoverOpen] = useState(false);
+  const [tableOptions, setTableOptions] = useState<EuiSelectableOption[]>(
+    TRACE_TABLE_OPTIONS.map((obj) =>
+      obj.key === props.tracesTableMode ? { ...obj, checked: 'on' } : obj
+    )
+  );
```
Comment (joshuali925) [inline public/components/trace_analytics/components/traces/traces_custom_indices_table.tsx:60]:
```suggestion
  const [tableOptions, setTableOptions] = useState<EuiSelectableOption[]>(
    () => TRACE_TABLE_OPTIONS.map((obj) =>
      obj.key === props.tracesTableMode ? { ...obj, checked: 'on' } : obj
    )
  );
```
compute initial value once should be enough

**Review by joshuali925 — APPROVED** [review]

File: public/components/trace_analytics/components/traces/traces_custom_indices_table.tsx (hunk 4 above)
Comment (ps48) [reply public/components/trace_analytics/components/traces/traces_custom_indices_table.tsx]: updated thanks to lazy init.

File: public/components/trace_analytics/index.scss (hunk 3 above)
Comment (ps48) [reply public/components/trace_analytics/index.scss:123]: OUI had `!important` on parent, but the selector works without important so removed it. Thanks!

File: public/components/trace_analytics/components/traces/trace_table_helpers.tsx (hunk 2 above)
Comment (ps48) [reply public/components/trace_analytics/components/traces/trace_table_helpers.tsx]: makes sense, removed lodash trucate usage in this file.

File: public/components/trace_analytics/requests/queries/traces_queries.ts [hunk 5]
```diff
@@ -557,20 +562,28 @@ export const getCustomIndicesTracesQuery = (
...
         should: [],
         must_not: [],
       },
     },
     ...(sort && { sort: [{ [sort.field]: { order: sort.direction } }] }),
     track_total_hits: false,
   };
+
+  if (queryMode === 'trace_root_spans') {
+    dataPrepperQuery.query.bool.filter.push({
```
Comment (vamsi-amazon) [inline public/components/trace_analytics/requests/queries/traces_queries.ts:575]: Not Related: we should have used an opensource name.

**Review by vamsi-amazon — APPROVED** [review]

File: common/constants/trace_analytics.ts [hunk 6]
```diff
@@ -24,3 +24,35 @@ export const TRACE_ANALYTICS_DSL_ROUTE = '/api/observability/trace_analytics/que
...
+  },
+  {
+    label: TRACE_TABLE_TITLES.trace_root_spans,
+    key: 'trace_root_spans',
+    'aria-describedby': 'The root spans which represent the starting point of a trace',
+  },
+  {
+    label: TRACE_TABLE_TITLES.service_entry_spans,
+    key: 'service_entry_spans',
+    'aria-describedby': 'The spans that mark start of server-side processing (SPAN_KIND_SERVER)',
```
Comment (vamsi-amazon) [inline common/constants/trace_analytics.ts:49]: I feel SPAN_KIND_SERVER is not needed. If required, bring in the exact description of SPAN_KIND_SERVER from open telemetry.

Comment (vamsi-amazon) [reply common/constants/trace_analytics.ts:49]: How is one and four different?

File: common/constants/trace_analytics.ts [hunk 7]
```diff
@@ -24,3 +24,35 @@ export const TRACE_ANALYTICS_DSL_ROUTE = '/api/observability/trace_analytics/que
...
+  },
+  {
+    label: TRACE_TABLE_TITLES.traces,
+    key: 'traces',
+    'aria-describedby': 'Aggregates all spans by traceId to show all traces',
```
Comment (vamsi-amazon) [inline common/constants/trace_analytics.ts:54]: `Spans aggregated by traceId to show all traces`
 Maintaining consistency where we start all the descriptions by `Spans`.
 
 Do we need to show all traces?
 
 How is this different from first option?

File: common/constants/trace_analytics.ts [hunk 8]
```diff
@@ -24,3 +24,35 @@ export const TRACE_ANALYTICS_DSL_ROUTE = '/api/observability/trace_analytics/que
...
 export const TRACE_CUSTOM_SERVICE_INDEX_SETTING = 'observability:traceAnalyticsServiceIndices';
+
+export enum TRACE_TABLE_TITLES {
+  all_spans = 'Spans',
+  trace_root_spans = 'Trace root spans',
+  service_entry_spans = 'Service entry spans',
+  traces = 'Traces',
+}
+
+export const TRACE_TABLE_OPTIONS = [
```
Comment (vamsi-amazon) [inline common/constants/trace_analytics.ts:35]: Users might ask for caching selected option.

File: common/constants/trace_analytics.ts [hunk 9]
```diff
@@ -24,3 +24,35 @@ export const TRACE_ANALYTICS_DSL_ROUTE = '/api/observability/trace_analytics/que
...
+  {
+    label: TRACE_TABLE_TITLES.all_spans,
+    key: 'all_spans',
+    'aria-describedby': 'All spans from all traces',
```
Comment (vamsi-amazon) [inline common/constants/trace_analytics.ts:39]: `Span from all traces.`
??

File: common/constants/trace_analytics.ts [hunk 10]
```diff
@@ -24,3 +24,35 @@ export const TRACE_ANALYTICS_DSL_ROUTE = '/api/observability/trace_analytics/que
...
+  },
+  {
+    label: TRACE_TABLE_TITLES.trace_root_spans,
+    key: 'trace_root_spans',
+    'aria-describedby': 'The root spans which represent the starting point of a trace',
```
Comment (vamsi-amazon) [inline common/constants/trace_analytics.ts:44]: `Root spans representing the starting point of each trace`

File: common/constants/trace_analytics.ts (hunk 6 above)
Comment (vamsi-amazon) [reply common/constants/trace_analytics.ts:49]: `Spans marking the start of server-side processing`
??

File: common/constants/trace_analytics.ts (hunk 8 above)
Comment (ps48) [reply common/constants/trace_analytics.ts:35]: yes, for now I've added the selection in the session storage similar to how we store filters. So the table type selection is persisted in the session lifecycle.

File: common/constants/trace_analytics.ts (hunk 7 above)
Comment (ps48) [reply common/constants/trace_analytics.ts:54]: This table is the old way of showing just traces using the aggregation query. Kept this option if users still want aggregation and are not concerned about the wait time.

Comment (ps48) [reply common/constants/trace_analytics.ts:54]: This is the only table where we shows traces and not spans

**Review by RyanL1997 — APPROVED** [review]
In generally LGTM!

File: public/components/trace_analytics/requests/queries/traces_queries.ts (hunk 5 above)
Comment (ps48) [reply public/components/trace_analytics/requests/queries/traces_queries.ts:575]: updated to `root_spans`

File: common/constants/trace_analytics.ts (hunk 6 above)
Comment (ps48) [reply common/constants/trace_analytics.ts:49]: updated to `Spans that indicate the entry point of service-side processing`

File: common/constants/trace_analytics.ts (hunk 7 above)
Comment (ps48) [reply common/constants/trace_analytics.ts:54]: updated: `Spans grouped by traceId to show a complete trace lifecycle`

File: common/constants/trace_analytics.ts (hunk 9 above)
Comment (ps48) [reply common/constants/trace_analytics.ts:39]: Updated: `spans representing all activities in all traces across the system`

**Pull request merged**